File: services/server/src/safe_socket/safe_socket.py
import socket

# recv_all reads exactly `size` bytes from socket, looping to handle
# short reads. Raises ConnectionError if the connection is closed
# before all bytes could be received.
def recv_all(sock: socket.socket, size):
    accumulated = 0
    buffer = bytearray()
    while accumulated < size:
        chunk = sock.recv(size - accumulated)
        if len(chunk) == 0:
            raise ConnectionError("Socket connection closed before receiving all bytes")
        buffer += chunk
        accumulated += len(chunk)
    return bytes(buffer)


# send_all writes all bytes to socket, looping to handle short writes.
# Raises ConnectionError if the connection is closed before all bytes
# could be sent.
def send_all(sock: socket.socket, data):
    accumulated = 0
    buffer_size = len(data)
    while accumulated < buffer_size:
        n = sock.send(data[accumulated:])
        if n == 0:
            raise ConnectionError("Socket connection closed before sending all bytes")
        accumulated += n

File: services/client/src/safe_socket/safe_socket.go
package safe_socket

import "io"


// SendAll writes all bytes to socket, looping to handle short writes.
// Returns an error if the write fails or the connection is closed
// before all bytes could be sent.
func SendAll(socket io.Writer, bytes []byte) error {
	accumulated := 0
	bufferSize := len(bytes)
	for accumulated < bufferSize {
		n, err := socket.Write(bytes[accumulated:])
		if err != nil {
			return err
		}
		accumulated += n
	}
	return nil
}

// RecvAll reads all bytes from socket, looping to handle short reads.
// Returns an error if the read fails or the connection is closed
// before all bytes could be received.
func RecvAll(socket io.Reader, size int) ([]byte, error) {
	accumulated := 0
	buffer := make([]byte, size)
	for accumulated < size {
		n, err := socket.Read(buffer[accumulated:])
		if err != nil {
			return nil, err
		}
		accumulated += n
	}
	return buffer, nil
}
